### PR TITLE
Support for incremental update of materialized sampletype

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
We can avoid needing to recreate the materialized sampletype temp table by handling some common cases (insert,delete,rollup).

#### Related Pull Requests
- 

#### Changes
-

### CODE REVIEW

File: experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java [hunk 1]
```diff
@@ -1069,19 +1110,186 @@ private SQLFragment getMaterializedSQL()
...
+                if (!materialized.incrementalRollupCheck.stillValid(0))
+                    executeIncrementalRollup();
+                if (!materialized.incrementalInsertCheck.stillValid(0))
+                    executeIncrementalInsert();
+            }
+            catch (RuntimeException|InterruptedException ex)
+            {
+                RuntimeException rex = UnexpectedException.wrap(ex);
+                materialized.setError(rex);
+                // The only time I'd expect and error is due to a schema change race-condition, but that can happen in any code path.
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java]:
```suggestion
                // The only time I'd expect an error is due to a schema change race-condition, but that can happen in any code path.

```

**Review by labkey-susanh — APPROVED** [review]

File: api/src/org/labkey/api/cache/BlockingCache.java [hunk 2]
```diff
@@ -91,6 +91,24 @@ public V get(@NotNull K key, @Nullable Object argument)
     }
 
 
+    public V getIfCached(@NotNull K key)
```
Comment (labkey-nicka) [inline api/src/org/labkey/api/cache/BlockingCache.java]: Consider adding test coverage in `BlockingCacheTest`.

nit: This appears to be unused code.

File: experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java [hunk 3]
```diff
@@ -1069,19 +1110,187 @@ private SQLFragment getMaterializedSQL()
...
+        {
+            DbSchema temp = DbSchema.getTemp();
+            String name = _prefix + "_" + GUID.makeHash();
+            _Materialized materialized = new _Materialized(this, name, txCacheKey, HeartBeat.currentTimeMillis(), "\"" + temp.getName() + "\".\"" + name + "\"");
+            initMaterialized(materialized);
+            return materialized;
+        }
+
+        @Override
+        public SQLFragment getFromSql(@NotNull SQLFragment selectQuery, boolean isSelectInto, String tableAlias)
```
Comment (labkey-nicka) [inline experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java]: nit: Remove this is a do nothing override.

**Review by labkey-nicka — APPROVED** [review]

File: api/src/org/labkey/api/cache/BlockingCache.java (hunk 2 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/cache/BlockingCache.java]: Good catch.  I think that went away when I moved from immediate to delayed update.

**Pull request merged**